hollow/src/main/java/com/netflix/hollow/tools/history/HollowHistory.java

```diff
@@ -354,8 +552,29 @@ private void addHistoricalState(HollowHistoricalState historicalState) {
         }
     }
 
+    // place historicalState at the end of historicalStates
+    //
+    // historicalStates is ordered like: V3 -> V2 -> V1
+    // however internally the states are linked like: V1.nextState = V2; V2.nextState = V3; etc.
+    private void addReverseHistoricalState(HollowHistoricalState historicalState) {
+        if (historicalStates.size() > 0) {
+            historicalState.getDataAccess().setNextState(historicalStates.get(historicalStates.size()-1).getDataAccess());
+            historicalState.setNextState(historicalStates.get(historicalStates.size()-1));
+        } else { // if reverse delta occurs before any fwd deltas
+            historicalState.getDataAccess().setNextState(latestHollowReadStateEngine);
+        }
+
+        historicalStates.add(historicalState);
+        historicalStateLookupMap.put(historicalState.getVersion(), historicalState);
+
+        if (historicalStates.size() >= maxHistoricalStatesToKeep) {
+            // drop old read state because we won't be building history in reverse after we get here
+            oldestHollowReadStateEngine = null;
```

Comment:
Ah, just learned that `HollowHistoricalState`s hold some references into the typestates contained within `oldestHollowHistoryState` so dereferencing it here won't actually free up much memory. I think we can drop those references once the `HollowHistoricalState` has been populated. Adding that...

Reply:
Added code to dereference read state engine once HollowHistoricalState has been populated so it can be GC'ed.